gsrs-module-substances-core/src/main/java/ix/ginas/models/v1/Substance.java

```diff
@@ -1496,4 +1498,77 @@ public static String toHexString(byte[] bytes) {
 
         return hexString.toString();
     }
+    
+    
+    // tags begin
+
+    public List<String> grabTagTerms() {
+        // This was named "getTagTerms," but I got strange PojoDiff error.
+        // It was a 500 error complaining about equal index and size values
+        // in the list. It occurred when trying to delete a tag *and* persisting.
+        List tagTerms = new ArrayList<String>();
+        for(Keyword k: this.tags){
+            tagTerms.add(k.getValue());
+        }
+        return tagTerms;
+    }
+
+    public Set<String> extractTagTermsFromNamesAsSet() {
+        List<Name> names = this.getAllNames();
+        if (names == null) { return null; }
+        Set<String> tagTerms = new HashSet<String>();
+        for (Name name: names) {
+            String e = name.extractTagTermFromName();
+            if (e != null) {
+                tagTerms.add(e);
+            }
+        }
+        return tagTerms;
+    }
+
+    public List<String> extractDistinctTagTermsFromNames() {
+        List<Name> names = this.getAllNames();
+        Set set = new HashSet<String>();
+        if (names == null) { return null; }
+        List tagTerms = new ArrayList<String>();
+        for (Name name: names) {
+            String e = name.extractTagTermFromName();
+            if (e != null) {
+                if (!set.contains(e)) {
+                    tagTerms.add(e);
+                    set.add(e);
+                }
+            }
+        }
+        return tagTerms;
+    }
+
+    public List<String> extractTagTermsFromNames() {
+        // Possibly returns non-distinct list.
+        List<Name> names = this.getAllNames();
+        if (names == null) { return null; }
+        List tagTerms = new ArrayList<String>();
+        for (Name name: names) {
+            String e = name.extractTagTermFromName();
+            if (e != null) {
+                tagTerms.add(e);
+            }
+        }
+        return tagTerms;
+    }
+
+    public List<String> compareTagTermsInNamesMissingFromTags(List<String> inNames, List<String> inTags) {
```

Comment:
Both of these methods below feel like Util methods. They're just doing basic set intersection things. They should be static utility methods in a Util class.